File: Problems/1992. Find All Groups of Farmland/find-all-groups-of-farmland.java
//Solution 1: 6ms 59.49 MB
class Solution {
  private int nr;
  private int nc;

  private void clearFarmland(
      int row, int col, int[][] land, int[] rightBottom) {
    if (row < 0 || row >= nr || col < 0 || col >= nc || land[row][col] == 0)
      return;

    land[row][col] = 0;
    rightBottom[0] = Math.max(rightBottom[0], row);
    rightBottom[1] = Math.max(rightBottom[1], col);

    clearFarmland(row + 1, col, land, rightBottom);
    clearFarmland(row, col + 1, land, rightBottom);
  }

  public int[][] findFarmland(int[][] land) {
    nr = land.length;
    nc = land[0].length;
    ArrayList<int[]> farmlandsList = new ArrayList<>();

    for (int i = 0; i < nr; i++) {
      for (int j = 0; j < nc; j++) {
        if (land[i][j] == 1) {
          int[] rightBottom = new int[2];
          rightBottom[0] = i;
          rightBottom[1] = j;
          clearFarmland(i, j, land, rightBottom);
          farmlandsList.add(new int[] {i, j, rightBottom[0], rightBottom[1]});
        }
      }
    }

    int[][] farmlands = new int[farmlandsList.size()][];
    for (int i = 0; i < farmlandsList.size(); i++) {
      farmlands[i] = farmlandsList.get(i);
    }

    return farmlands;
  }
}

//Solution 2: 5ms 59.91 MB
class Solution {
  private int nr;
  private int nc;

  private int rbr;
  private int rbc;

  private void clearFarmland(int row, int col, int[][] land) {
    if (row < 0 || row >= nr || col < 0 || col >= nc || land[row][col] == 0) return;

    land[row][col] = 0;
    rbr = Math.max(rbr, row);
    rbc = Math.max(rbc, col);

    clearFarmland(row + 1, col, land);
    clearFarmland(row, col + 1, land);
  }

  public int[][] findFarmland(int[][] land) {
    nr = land.length;
    nc = land[0].length;
    ArrayList<int[]> farmlandsList = new ArrayList<>();

    for (int i = 0; i < nr; i++) {
      for (int j = 0; j < nc; j++) {
        if (land[i][j] == 1) {
          rbr = i;
          rbc = j;
          clearFarmland(i, j, land);
          farmlandsList.add(new int[] {i, j, rbr, rbc});
        }
      }
    }

    int[][] farmlands = new int[farmlandsList.size()][];
    for (int i = 0; i < farmlandsList.size(); i++) {
      farmlands[i] = farmlandsList.get(i);
    }

    return farmlands;
  }
}

//Solution 3: 4ms 59.73 MB
class Solution {
  private int nr;
  private int nc;

  private int rbr;
  private int rbc;

  private void clearFarmland(int row, int col, int[][] land) {
    if (row >= nr || col >= nc || land[row][col] == 0) return;

    land[row][col] = 0;
    rbr = Math.max(rbr, row);
    rbc = Math.max(rbc, col);

    clearFarmland(row + 1, col, land);
    clearFarmland(row, col + 1, land);
  }

  public int[][] findFarmland(int[][] land) {
    nr = land.length;
    nc = land[0].length;
    ArrayList<int[]> farmlandsList = new ArrayList<>();

    for (int i = 0; i < nr; i++) {
      for (int j = 0; j < nc; j++) {
        if (land[i][j] == 1) {
          rbr = i;
          rbc = j;
          clearFarmland(i, j, land);
          farmlandsList.add(new int[] {i, j, rbr, rbc});
        }
      }
    }

    int[][] farmlands = new int[farmlandsList.size()][];
    for (int i = 0; i < farmlandsList.size(); i++) {
      farmlands[i] = farmlandsList.get(i);
    }

    return farmlands;
  }
}

//Solution 4: 3ms 60.42 MB
class Solution {
  private int nr;
  private int nc;

  private int[] findFarmlandCoordinates(int row, int col, int[][] land) {
    int[] coordinates = new int[4];
    coordinates[0] = row;
    coordinates[1] = col;
    int rbr = row;
    int rbc = col;

    while (rbr < nr && land[rbr][col] == 1) rbr++;
    while (rbc < nc && land[row][rbc] == 1) rbc++;

    coordinates[2] = rbr - 1;
    coordinates[3] = rbc - 1;

    for (int i = row; i < rbr; i++) {
      for (int j = col; j < rbc; j++) {
        land[i][j] = 0;
      }
    }

    return coordinates;
  }

  public int[][] findFarmland(int[][] land) {
    nr = land.length;
    nc = land[0].length;
    ArrayList<int[]> farmlandsList = new ArrayList<>();

    for (int i = 0; i < nr; i++) {
      for (int j = 0; j < nc; j++) {
        if (land[i][j] == 1) {
          farmlandsList.add(findFarmlandCoordinates(i, j, land));
        }
      }
    }

    int[][] farmlands = new int[farmlandsList.size()][];
    for (int i = 0; i < farmlandsList.size(); i++) {
      farmlands[i] = farmlandsList.get(i);
    }

    return farmlands;
  }
}

File: Problems/1992. Find All Groups of Farmland/find-all-groups-of-farmland.ts
//Solution 1: 253ms 84.70 MB
function findFarmland(land: number[][]): number[][] {
  const landRowLength = land.length;
  const landColLength = land[0].length;
  let topLeft: [number, number] = [-1, -1];
  let bottomRight: [number, number] = [-1, -1];
  function clearFarmland(row: number, col: number) {
    if (row >= landRowLength || col >= landColLength || land[row][col] === 0)
      return;
    bottomRight[0] = Math.max(bottomRight[0], row);
    bottomRight[1] = Math.max(bottomRight[1], col);
    land[row][col] = 0;
    clearFarmland(row + 1, col);
    clearFarmland(row, col + 1);
  }

  const answer = [];
  for (let i = 0; i < land.length; i++) {
    for (let j = 0; j < land[0].length; j++) {
      if (land[i][j] === 1) {
        topLeft[0] = i;
        topLeft[1] = j;
        bottomRight[0] = i;
        bottomRight[1] = j;
        clearFarmland(i, j);
        answer.push([...topLeft, ...bottomRight]);
      }
    }
  }

  return answer;
}

//Solution 2: 230ms 86.04 MB
let nr: number = 0;
let nc: number = 0;

function findFarmlandCoordinates(
  row: number,
  col: number,
  land: number[][]
): number[] {
  let coordinates: number[] = [row, col];
  let rbr: number = row;
  let rbc: number = col;

  while (rbr < nr && land[rbr][col] === 1) rbr++;
  while (rbc < nc && land[row][rbc] === 1) rbc++;

  coordinates[2] = rbr - 1;
  coordinates[3] = rbc - 1;

  for (let i = row; i < rbr; i++) {
    for (let j = col; j < rbc; j++) {
      land[i][j] = 0;
    }
  }

  return coordinates;
}

function findFarmland(land: number[][]): number[][] {
  nr = land.length;
  nc = land[0].length;
  let farmlandsList: number[][] = [];

  for (let i = 0; i < nr; i++) {
    for (let j = 0; j < nc; j++) {
      if (land[i][j] === 1) {
        farmlandsList.push(findFarmlandCoordinates(i, j, land));
      }
    }
  }

  return farmlandsList;
}

//Solution 3: 183ms 75.02 MB
function findFarmland(land: number[][]): number[][] {
  const rows = land.length;
  const cols = land[0].length;
  let bottomRightX = -1;
  let bottomRightY = -1;

  function clearFarmland(row: number, col: number) {
    if (row >= rows || col >= cols || land[row][col] === 0) return;
    bottomRightX = Math.max(bottomRightX, row);
    bottomRightY = Math.max(bottomRightY, col);
    land[row][col] = 0;
    clearFarmland(row + 1, col);
    clearFarmland(row, col + 1);
  }

  const farmlands: number[][] = [];
  for (let topLeftX = 0; topLeftX < rows; topLeftX++) {
    for (let topLeftY = 0; topLeftY < cols; topLeftY++) {
      if (land[topLeftX][topLeftY] === 1) {
        bottomRightX = topLeftX;
        bottomRightY = topLeftY;
        clearFarmland(topLeftX, topLeftY);
        farmlands.push([topLeftX, topLeftY, bottomRightX, bottomRightY]);
      }
    }
  }

  return farmlands;
}
